pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go

```diff
@@ -299,7 +299,59 @@ func revertToCutoverTimestamp(
 // TODO(adityamaru): Add ClearRange logic once we have introduced
 // synchronization between the flow tearing down and the job transitioning to a
 // failed/canceled state.
-func (s *streamIngestionResumer) OnFailOrCancel(_ context.Context, _ interface{}) error {
+func (s *streamIngestionResumer) OnFailOrCancel(ctx context.Context, execCtx interface{}) error {
+	jobExecCtx := execCtx.(sql.JobExecContext)
+	db := jobExecCtx.ExecCfg().DB
+	p := s.job.Payload()
+	spans := []roachpb.Span{p.GetStreamIngestion().Span}
+	details := s.job.Details().(jobspb.StreamIngestionDetails)
+	client, err := streamclient.NewStreamClient(streamingccl.StreamAddress(details.StreamAddress))
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Cancel the producer job even when the revert has error. The source job's
+		// protected timestamp is no longer needed as this ingestion job won't resume
+		// ingestion anymore.
+		err := client.Complete(ctx, streaming.StreamID(details.StreamID), false /* ingestionCutover*/)
+		if err != nil {
+			log.Warningf(ctx, "encountered error when canceling the producer job: %v", err)
+		}
+		err = client.Close()
+		if err != nil {
+			log.Warningf(ctx, "encountered error when closing the stream client: %v", err)
+		}
+	}()
+	for len(spans) != 0 {
+		var b kv.Batch
+		for _, span := range spans {
+			b.AddRawRequest(&roachpb.ClearRangeRequest{
+				RequestHeader: roachpb.RequestHeader{
+					Key:    span.Key,
+					EndKey: span.EndKey,
+				},
+			})
+		}
+		if err := db.Run(ctx, &b); err != nil {
+			return err
+		}
+
+		// TODO(casper): handle timeout when we have very large tenant with
+		// saving checkpoints and retrying.
```

Comment:
I'm not sure about timeout (we haven't set it have we? And the context doesn't have a deadline set either); but I do think we need to add checks to ensure the requirements specified by the ClearRange request: namely, we need to ensure that the spans are offline.
https://github.com/cockroachdb/cockroach/blob/master/pkg/roachpb/api.proto#L384-L384